custom/custompfm/keeper/keeper.go

```diff
@@ -0,0 +1,311 @@
+package keeper
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/armon/go-metrics"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+	router "github.com/cosmos/ibc-apps/middleware/packet-forward-middleware/v7/packetforward"
+	"github.com/cosmos/ibc-apps/middleware/packet-forward-middleware/v7/packetforward/keeper"
+	"github.com/cosmos/ibc-apps/middleware/packet-forward-middleware/v7/packetforward/types"
+	transfertypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v7/modules/core/05-port/types"
+	ibcexported "github.com/cosmos/ibc-go/v7/modules/core/exported"
+	custombankkeeper "github.com/notional-labs/composable/v6/custom/bank/keeper"
+	ibctransfermiddlewarekeeper "github.com/notional-labs/composable/v6/x/ibctransfermiddleware/keeper"
+)
+
+var _ porttypes.Middleware = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the forward middleware given the
+// forward keeper and the underlying application.
+type IBCMiddleware struct {
+	router.IBCMiddleware
+
+	app1    porttypes.IBCModule
+	keeper1 *keeper.Keeper
+
+	retriesOnTimeout1 uint8
+	forwardTimeout1   time.Duration
+	refundTimeout1    time.Duration
+	ibcfeekeeper      *ibctransfermiddlewarekeeper.Keeper
+	bank              *custombankkeeper.Keeper
+}
+
+func NewIBCMiddleware(
+	app porttypes.IBCModule,
+	k *keeper.Keeper,
+	retriesOnTimeout uint8,
+	forwardTimeout time.Duration,
+	refundTimeout time.Duration,
+	ibcfeekeeper *ibctransfermiddlewarekeeper.Keeper,
+	bankkeeper *custombankkeeper.Keeper,
+) IBCMiddleware {
+	return IBCMiddleware{
+		IBCMiddleware: router.NewIBCMiddleware(app, k, retriesOnTimeout, forwardTimeout, refundTimeout),
+		ibcfeekeeper:  ibcfeekeeper,
+
+		app1:              app,
+		keeper1:           k,
+		retriesOnTimeout1: retriesOnTimeout,
+		forwardTimeout1:   forwardTimeout,
+		refundTimeout1:    refundTimeout,
+		bank:              bankkeeper,
+	}
+}
+
+func (im IBCMiddleware) OnRecvPacket(
```

Comment:
don't you need to implement all of the methods?

Reply:
no, all other methods are implemented because i include into mine IbcMidleware type the original IbcMiddleware that bring implementation of others methods as inherited from original
I just ovveride the only one method to inject my custom fee logic